shared/actions/__tests__/signup.test.js

```diff
@@ -0,0 +1,393 @@
+// @flow
+/* eslint-env jest */
+import * as Constants from '../../constants/signup'
+import * as SignupGen from '../signup-gen'
+import * as LoginGen from '../login-gen'
+import * as Saga from '../../util/saga'
+import type {TypedState} from '../../constants/reducer'
+import {loginTab} from '../../constants/tabs'
+import HiddenString from '../../util/hidden-string'
+import {navigateUp, navigateTo, navigateAppend} from '../route-tree'
+import {_testing} from '../signup'
+import reducer from '../../reducers/signup'
+
+jest.unmock('immutable')
+
+describe('resetNav works', () => {
+  it('nab based on login', () => {
+    expect(_testing.resetNav()).toEqual(Saga.put(LoginGen.createNavBasedOnLoginAndInitialState()))
+  })
+})
+
+describe('goBackAndClearErrors', () => {
+  it('errors get cleaned and we go back a level', () => {
+    const getState = () => ({
+      signup: Constants.makeState({
+        devicenameError: 'bad name',
+        emailError: 'bad email',
+        inviteCodeError: 'bad invite',
+        nameError: 'bad name',
+        passphraseError: new HiddenString('bad pass'),
+        signupError: new HiddenString('bad signup'),
+        usernameError: 'bad username',
+      }),
+    })
+
+    const action = SignupGen.createGoBackAndClearErrors()
+    const nextState = {signup: reducer(getState().signup, action)}
+    expect(nextState.signup.devicenameError).toEqual('')
+    expect(nextState.signup.emailError).toEqual('')
+    expect(nextState.signup.inviteCodeError).toEqual('')
+    expect(nextState.signup.nameError).toEqual('')
+    expect(nextState.signup.passphraseError.stringValue()).toEqual('')
+    expect(nextState.signup.signupError.stringValue()).toEqual('')
+    expect(nextState.signup.usernameError).toEqual('')
+    expect(_testing.goBackAndClearErrors()).toEqual(Saga.put(navigateUp()))
+  })
+})
+
+describe('requestAutoInvite', () => {
+  const getState = () => ({signup: Constants.makeState({})})
```

Comment:
Most of the test code just works with the `signup` field, so you can save a bunch of boilerplate by just working with that directly:

```
  it('makes a call to get an auto invite', () => {
    const action = SignupGen.createRequestAutoInvite()
    const nextState = reducer(Constants.makeState(), action)
    expect(nextState).toEqual(Constants.makeState())
  })
```

The only place you do need the full state is with the various `doXOnNoErrors` helpers, but you can probably just make a `signupStateToFullState` helper for that:

```
  it('ignores if theres an error', () => {
    const action = SignupGen.createRequestInvite({email: 'email@email.com', name: 'name'})
    const nextState = reducer(Constants.makeState(), action)
    expect(_testing.showInviteSuccessOnNoErrors(signupStateToFullState(nextState))).toEqual(false)
  })
```
or refactor the various `doXOnNoErrors` functions to take the signup state instead of the full state, and pass some wrappers to `safeTakeEveryPureSimple`

Reply:
yeah i had thought about it but i was making it more generic in case you wanted to have actions that touch the whole state and it'd be easier to copy / paste around but i guess thats rare. i'll remove it